accounting/invoice/tasks.py

```diff
@@ -0,0 +1,94 @@
+# -*- coding: utf-8 -*-
+#
+# OpenCraft -- tools to aid developing and hosting free software projects
+# Copyright (C) 2017-2018 OpenCraft <contact@opencraft.com>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+"""
+Invoice tasks.
+"""
+
+from urllib.parse import urljoin
+import datetime
+
+from django.conf import settings
+from django.contrib.sites.models import Site
+from django.core.mail.message import EmailMessage
+from django.template.loader import get_template
+from huey import crontab
+from huey.contrib.djhuey import db_periodic_task
+from rest_framework.reverse import reverse
+
+from accounting.account.models import Account
+from accounting.invoice import constants
+from accounting.invoice.utils import send_email_with_invoice
+
+
+@db_periodic_task(crontab(day=settings.INVOICE_NOTIFICATION_DAY, hour="0", minute="0"))
+def send_invoice_prep_notification():
+    """
+    Send emails to notify invoice recipients to prepare for invoice receipt.
+    """
+    now = datetime.datetime.now()
+    approval_date = now.replace(day=int(settings.INVOICE_APPROVAL_DAY))
+    past_month = (now - datetime.timedelta(days=now.day)).strftime('%B')
+    template = get_template(constants.INVOICE_NOTIFICATION_TEMPLATE)
+    for client_username in settings.BILLING_CYCLE_USERS:
+        client = Account.objects.get(user__username=client_username)
+        provider_emails = [rate.provider.user.email for rate in client.client_hourly_rates.all()]
+        message = template.render({'approval_date': approval_date, 'month': past_month})
+        EmailMessage(
+            subject=constants.INVOICE_NOTIFICATION_SUBJECT.format(month=past_month),
+            body=message,
+            bcc=provider_emails,
```

Comment:
Well actually in this case, there's no invoice being attached, because this is just for the original notification to prepare for the invoice on the 3rd. Sending a single email with an attached file that should differ per recipient isn't possible.

Django does have a `send_mass_email` function, which opens a single connection to send emails to multiple `to` targets, but it ends up including all of their names in the list, which is something I wanted to hide, with `bcc`.

Reply:
@UmanShahzad Man... I'm a little tired today! What I meant to say was:

> This makes it possible to send a single e-mail and still have each provider receive only their own invoice **notification**

Gah! :)

Reply:
All good :+1: